From ebcd1abb56313e07bb786246cb04b28b43c2ac96 Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Wed, 9 Jul 2014 14:21:58 +0100
Subject: [PATCH 263/351] ASoC: arizona: Mute all analog channels whilst
 clearing DAC

Mute all analogue channels whilst we are clearing the DAC after a
channel enable. This is to prevent pop noises appearing on the other
channels as the DAC clear affects all channels.

Change-Id: I0cdbc828f12e52d63fbddb592fdeb42eddc93ee3
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
---
 drivers/mfd/arizona-core.c       |    5 +++--
 include/linux/mfd/arizona/core.h |    2 ++
 sound/soc/codecs/arizona.c       |    6 ++++++
 3 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/mfd/arizona-core.c b/drivers/mfd/arizona-core.c
index 387b192..ce1be6e0 100644
--- a/drivers/mfd/arizona-core.c
+++ b/drivers/mfd/arizona-core.c
@@ -982,8 +982,8 @@ static const struct {
 	},
 };
 
-static void arizona_florida_mute_analog(struct arizona* arizona,
-					unsigned int mute)
+void arizona_florida_mute_analog(struct arizona* arizona,
+				 unsigned int mute)
 {
 	unsigned int val, chans;
 	int i;
@@ -1007,6 +1007,7 @@ static void arizona_florida_mute_analog(struct arizona* arizona,
 				   mute);
 	}
 }
+EXPORT_SYMBOL_GPL(arizona_florida_mute_analog);
 
 static bool arizona_florida_get_input_state(struct arizona* arizona)
 {
diff --git a/include/linux/mfd/arizona/core.h b/include/linux/mfd/arizona/core.h
index 88798b9..867c5cf 100644
--- a/include/linux/mfd/arizona/core.h
+++ b/include/linux/mfd/arizona/core.h
@@ -144,6 +144,8 @@ extern int arizona_of_read_u32_array(struct arizona *arizona, const char *prop,
 extern int arizona_of_read_u32(struct arizona *arizona, const char* prop,
 			       bool mandatory, u32 *data);
 
+extern void arizona_florida_mute_analog(struct arizona* arizona,
+					unsigned int mute);
 extern void arizona_florida_clear_input(struct arizona *arizona);
 
 #endif
diff --git a/sound/soc/codecs/arizona.c b/sound/soc/codecs/arizona.c
index 6077797..f2e02d7e 100644
--- a/sound/soc/codecs/arizona.c
+++ b/sound/soc/codecs/arizona.c
@@ -861,6 +861,10 @@ static int arizona_update_input(struct arizona* arizona, bool enable)
 	regmap_write(arizona->regmap, 0x80,  0x3);
 
 	if (enable) {
+		arizona_florida_mute_analog(arizona, ARIZONA_IN1L_MUTE);
+
+		msleep(10);
+
 		regmap_write(arizona->regmap, 0x3A6, 0x5555);
 		regmap_write(arizona->regmap, 0x3A5, 0x3);
 	} else {
@@ -871,6 +875,8 @@ static int arizona_update_input(struct arizona* arizona, bool enable)
 			regmap_write(arizona->regmap, 0x3A6, 0x0);
 			msleep(5);
 		}
+
+		arizona_florida_mute_analog(arizona, 0);
 	}
 
 	regmap_write(arizona->regmap, 0x80,  0x0);
-- 
1.7.9.5

